File: algorithms/algorithms-cplusplus/leetcode/Question_0820_Short_Encoding_of_Words.cc
#include <string>
#include <vector>

using namespace std;

struct TrieNode {
  vector<TrieNode*> children;
  int count;
};

class Solution {
 public:
  int minimumLengthEncoding(vector<string>& words) {
    TrieNode *root = new TrieNode(), *cur;
    root->children.resize(26);
    for (string& w : words) {
      cur = root;
      for (int i = w.size() - 1; i >= 0; i--) {
        if (!cur->children[w[i] - 'a']) {
          cur->children[w[i] - 'a'] = new TrieNode();
          (cur->children[w[i] - 'a'])->children.resize(26);
          cur->count++;
        }
        cur = cur->children[w[i] - 'a'];
      }
    }
    return help(root, 0);
  }

 private:
  int help(TrieNode *root, int len) {
    if (root->count == 0) {
      return len + 1;
    }
    int res = 0;
    for (TrieNode *cur : root->children) {
      if (cur) {
        res += help(cur, len + 1);
      }
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0820_Short_Encoding_of_Words.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0820_Short_Encoding_of_Words {
    static class TrieNode {
        TrieNode[] children;
        int count;

        TrieNode() {
            children = new TrieNode[26];
            count = 0;
        }

        public TrieNode get(char c) {
            if (children[c - 'a'] == null) {
                children[c - 'a'] = new TrieNode();
                count++;
            }
            return children[c - 'a'];
        }
    }

    public int minimumLengthEncoding(String[] words) {
        TrieNode root = new TrieNode(), cur;
        for (String word : words) {
            cur = root;
            for (int j = word.length() - 1; j >= 0; --j) {
                cur = cur.get(word.charAt(j));
            }
        }
        return help(root, 0);
    }

    private int help(TrieNode root, int len) {
        if (root.count == 0) {
            return len + 1;
        }
        int res = 0;
        for (TrieNode p : root.children) {
            if (p != null) {
                res += help(p, len + 1);
            }
        }
        return res;
    }
}
